source/extensions/certificate_providers/default_cert_provider/config.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+#include "envoy/config/core/v3/extension.pb.h"
+#include "envoy/extensions/certificate_providers/default_cert_provider/v3/config.pb.h"
+
+#include "source/common/common/callback_impl.h"
+#include "source/extensions/certificate_providers/factory.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace CertificateProviders {
+
+class DefaultCertificateProvider : public CertificateProvider::CertificateProvider,
+                                  public CertificateProvider::CertificateSubscriptionCallbacks {
```

Comment:
It sounds to me like we may be conflating two different things here:

1. A mechanism used by a `CertificateProvider` implementation to subscribe to whatever data source it's getting the certificates from.
2. A mechanism used by clients of a `CertificateProvider` to subscribe to the certificates provided by the `CertificateProvider`.

It sounds like you intend `CertificateSubscriptionCallbacks` to be (1), not (2).  However, I don't think we actually need to define an API for (1), because every `CertificateProvider` implementation may use a different mechanism for obtaining the certificates (e.g., it may subscribe to an xDS resource, watch a file, contact an external control plane using some custom non-xDS protocol, or something we can't even think of right now).  I think we can let each `CertificateProvider` implementation handle that part however they want to; there is no benefit to trying to force implementations to use a provided interface like `CertificateSubscriptionCallbacks`, since that semantics of that class may not match the mechanism that they're actually using to obtain the certificates.

In contrast, I think we do need to define an API for (2).  The current `addUpdateCallback()` mechanism is not sufficient, because it does not allow callers of a `CertificateProvider` to unsubscribe to a certificate that they had previously subscribed to.  That functionality will be needed when the xDS config changes and the caller no longer needs to subscribe to the cert that it was using before.

Please re-read my [earlier comment](https://github.com/envoyproxy/envoy/pull/19582/files#r831270521) for a description of how I think this should be structured.

Reply:
> 1. A mechanism used by a `CertificateProvider` implementation to subscribe to whatever data source it's getting the certificates from.
> 2. A mechanism used by clients of a `CertificateProvider` to subscribe to the certificates provided by the `CertificateProvider`.

I was indeed to intend `CertificateSubscriptionCallbacks` to be (1),. Although each `CertificateProvider` has its own mechanism for obtaining the certificates, they can follow unified interfaces, this is what I was thinking about. I can remove it and let each `CertificateProvider` implementation handle that part in any free style.

> * A `CertificateProvider` instance will be instantiated for each entry in the list of plugins defined in the bootstrap file.  There will be a global method to obtain a given provider by name.
> * The `CertificateProvider` class will have a `startWatch()` method that takes an instance of `CertificateSubscriptionCallbacks` as a parameter, and another `cancelWatch()` method that allows cancelling a previously started watch.
> * Callers that are configured to use a `CertificateProvider` (like the upstream and downstream TLS handshaking code) will implement their own subclass of `CertificateSubscriptionCallbacks`.  They will use the global function to find the right named `CertificateProvider` and then call that provider's `startWatch()` method with an instance of their `CertificateSubscriptionCallbacks` implementation.

The name `startWatch` sounds it just makes the `CertificateProvider` to watch the updates of some cert. But I think you also want `startWatch` to register callback functions to `CertificateProvider` just like `addUpdateCallback` does. The `startWatch` can be contained in `addUpdateCallback`, we can make the `CertificateProvider` starts watch at the beginning of the `addUpdateCallback`.

> * Whenever there is one or more watcher for a given cert, the `CertificateProvider` will do whatever it needs to do to always have an up-to-date copy of that cert (e.g., watching a file or subscribing to an xDS resource).  This activity will start automatically when the first watch is started for a given cert, and it will stop automatically when the last watch for a given cert is cancelled.

Got it. So we need a `delUpdateCallback`, if we don't have any callback functions relevant to one cert, we cancel the watch on it.

> * Whenever the `CertificateProvider` obtains a new copy of a given cert, it will invoke the relevant method of all `CertificateSubscriptionCallbacks` objects that are watching that cert.  This notifies the caller that the cert has been updated.
> 
> I am not sure how you are intending this to work right now, but it seems very different to the above. I don't understand why it makes sense to have a single object implement both `CertificateProvider` and `CertificateSubscriptionCallbacks`.

In a word, I just need to add a "delUpdateCallback", right? Please correct me if I misunderstand anything.

Reply:
> I was indeed to intend `CertificateSubscriptionCallbacks` to be (1),. Although each `CertificateProvider` has its own mechanism for obtaining the certificates, they can follow unified interfaces, this is what I was thinking about. I can remove it and let each `CertificateProvider` implementation handle that part in any free style.

Yes, please do remove it.  I don't think we need that as part of this API.

> The name `startWatch` sounds it just makes the `CertificateProvider` to watch the updates of some cert. But I think you also want `startWatch` to register callback functions to `CertificateProvider` just like `addUpdateCallback` does. The `startWatch` can be contained in `addUpdateCallback`, we can make the `CertificateProvider` starts watch at the beginning of the `addUpdateCallback`.

Yes, `startWatch()` is intended to serve the same purpose for which you currently have `addUpdateCallback()`.

The name `startWatch()` is intended to express the fact that the caller is starting a watch on a particular cert with the `CertificateProvider`.  That may happen to cause the `CertificateProvider` to start a watch on some external resource, but that's just an implementation detail of the `CertificateProvider` implementation; it's completely irrelevant to the API that the `CertificateProvider` exports to its caller.  For example, you could have a `CertificateProvider` impl that just provides static cert data that it is given at instantiation time, in which case it does not need to ever start a watch of any external resource.  But that implementation detail would be completely hidden by the `CertificateProvider` API; the caller would just see a single update as soon as they start their watch on the `CertificateProvider` and then they'd never get another update.

I think we need more than just a single callback here.  You probably want to pass the certs to the watcher as part of notifying them of the update, or else you will have lock contention trying to call the accessor methods from multiple threads.  Also, we need some mechanism to notify the watcher when there is an error obtaining the resource, so that the watcher knows when it's waiting for a result (in which case the async TLS handshake can wait for the result) vs. when there was an error (in which case the handshake should be failed immediately).

You can see the API we use in gRPC [here](https://github.com/grpc/grpc/blob/44167bdfd3912a064696efc04403217d9b0e2314/src/core/lib/security/credentials/tls/grpc_tls_certificate_distributor.h#L41).

> Got it. So we need a `delUpdateCallback`, if we don't have any callback functions relevant to one cert, we cancel the watch on it.

Yes, exactly -- assuming that the `CertificateProvider` impl happens to use a watch on an external resource.  As mentioned above, that is entirely dependent on the `CertificateProvider` impl and should be completely hidden by this API.

> In a word, I just need to add a "delUpdateCallback", right? Please correct me if I misunderstand anything.

That's part of it.  But I think we also need to replace the callback with a watcher interface that has methods to deliver the certs and to indicate errors.

Reply:
Sorry for late response.

Well, I prefer the similar API like [secret provider](https://github.com/envoyproxy/envoy/blob/0fae6970ddaf93f024908ba304bbd2b34e997a51/envoy/secret/secret_provider.h#L18) in Envoy instead of the gPRC api. With secret provider the caller only needs registering a callback via [addUpdateCallback](https://github.com/envoyproxy/envoy/blob/bda2fb9cbf9b6ca9329da0f4c0e65cb3bb23f19e/source/extensions/transport_sockets/tls/context_config_impl.cc#L259), I'd like to make the certificate provider works similarly. The gPRC api requires the caller implement the watcher interface, including `OnCertificatesChanged` and `onError`, I think  `OnCertificatesChanged` function is similar to `addUpdateCallback`, and for `onError` the certificate provider could raise exception or output logs by itself, we don't need to require every caller to handle the update error.  Besides, `watcher` in grpc api sounds the interface the caller needs to implement,  while in Envoy there is `file Watcher` which is fundamental to implement the subscription to file data source. I don't want to make it looks confused.

After thinking for a while, I can add "delUpdateCallback", but when will the caller want to stop receiving the updates?

Reply:
@lizan @markdroth ping, please have a look at above message.

Reply:
@lizan @markdroth ping

Reply:
@markdroth Thanks, now I got why the `delUpdateCallback` is needed, but I think Envoy provides a delicate way to handle this case. We require the "addUpdateCallback" returns a [`CallbackHandlePtr`](https://github.com/envoyproxy/envoy/blob/96dd73500ab7910075a5810254234b41389fb813/source/common/common/callback_impl.h), it is a handler that can be used to remove the callback, see the code [destructor of CallbackHolder ](https://github.com/envoyproxy/envoy/blob/96dd73500ab7910075a5810254234b41389fb813/source/common/common/callback_impl.h#L82) , my understanding is LDS or CDS update will trigger old config objects destructor, so the callbacks will be removed during CallbackHandler destructor.

Reply:
I don't really care what the mechanism is for canceling the watch, as long as there's a way to do it.  If it's more idiomatic in Envoy to do it via something like `CallbackHandlePtr` rather than having a separate cancellation method, that's fine.

That having been said, I think this still needs to be a watcher interface instead of a simple callback, for the reasons I described earlier:
> I think we need more than just a single callback here. You probably want to pass the certs to the watcher as part of notifying them of the update, or else you will have lock contention trying to call the accessor methods from multiple threads. Also, we need some mechanism to notify the watcher when there is an error obtaining the resource, so that the watcher knows when it's waiting for a result (in which case the async TLS handshake can wait for the result) vs. when there was an error (in which case the handshake should be failed immediately).

Reply:
Not sure if you missed my reply before.

> Well, I prefer the similar API like [secret provider](https://github.com/envoyproxy/envoy/blob/0fae6970ddaf93f024908ba304bbd2b34e997a51/envoy/secret/secret_provider.h#L18) in Envoy instead of the gPRC-style api. With secret provider the caller only needs registering a callback via [addUpdateCallback](https://github.com/envoyproxy/envoy/blob/bda2fb9cbf9b6ca9329da0f4c0e65cb3bb23f19e/source/extensions/transport_sockets/tls/context_config_impl.cc#L259), I'd like to make the certificate provider works similarly. The gPRC api requires the caller implement the watcher interface, including `OnCertificatesChanged` and `onError`, I think  `OnCertificatesChanged` function is similar to `addUpdateCallback`,

we can add `onCertificatesChanged` callback function to Certificate Provider via `addUpdateCallback`, Certificate Provider will run all these callbacks when receiving certificates update which is implementation details. I don't think we need to call  the accessor methods from multiple threads.

> and for `onError` the certificate provider could raise exception or output logs by itself, we don't need to require every caller to handle the update error.  

similarly, `onError` in you grpc api is also a callback function, I think we don't need each caller to handle update error , if the caller really has its own solution for update error, the `addErrorCallbacks` in Certificate Provider might be a choice, when Certificate fails to update cert, we make it run all error callback functions.

> Besides, `watcher` in grpc api sounds the interface the caller needs to implement,  while in Envoy there is `file Watcher` which is fundamental to implement the subscription to file data source. I don't want to make it looks confused.

Does it solve your concern?

Reply:
@markdroth ping

Reply:
@markdroth @lizan ping

Reply:
@markdroth @ggreenway  you mentioned this work in https://github.com/envoyproxy/envoy/issues/21292

I'd like to change the name "certificate_provider_instances" to "secrect_provider_instances" first, because this is not just to provide certificates (cert and private key),  it should provide any kinds of [secret ](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/secret.proto#extensions-transport-sockets-tls-v3-secret) since it is equivalent to sds and will introduce more flexible capabilities. It also looks confusing during developing.

Does it make sense for you? Do you have other better name?